mrmustard/lab_dev/states/base.py

```diff
@@ -690,19 +683,31 @@ def from_phase_space(
         return ret
 
     @property
-    def probability(self) -> float:
+    def _probabilities(self) -> RealVector:
+        r"""Element-wise probabilities along the batch dimension of this DM.
+        Useful for cases where the batch dimension does not mean a convex combination of states."""
         idx_ket = self.wires.output.ket.indices
         idx_bra = self.wires.output.bra.indices
-
         rep = self.representation.trace(idx_ket, idx_bra)
-
         if isinstance(rep, Bargmann):
-            return math.real(math.sum(rep.c, axes=[0]))
-        return math.real(math.sum(rep.array, axes=[0]))
+            return math.real(rep.c)
+        return math.real(rep.array)
+
+    @property
+    def probability(self) -> float:
+        r"""Probability of this DM, using the batch dimension of the Ansatz
+        as a convex combination of states."""
+        return math.sum(self._probabilities)
+
+    @property
+    def _purities(self) -> RealVector:
```

Comment:
why do all these properties need to be private?

Reply:
they don't, I'll change it